x/gov/src/keeper.rs

```diff
@@ -314,9 +355,302 @@ impl<SK: StoreKey, PSK: ParamsSubspaceKey, M: Module, BK: BankKeeper<SK, M>>
+                            "proposal_id".into(),
+                            proposal.proposal_id.to_string().into(),
+                            false,
+                        ),
+                        EventAttribute::new("proposal_result".into(), "TODO".into(), false),
+                    ],
+                ))
+            }
+        }
+
+        events
+    }
+
+    fn tally<DB: Database, CTX: TransactionalContext<DB, SK>>(
+        &self,
+        ctx: &mut CTX,
+        proposal_id: u64,
+    ) -> Result<(bool, bool, TallyResult), GasStoreErrors> {
+        let mut curr_validators = HashMap::<ValAddress, ValidatorGovInfo>::new();
+
+        for validator in self.staking_keeper.bonded_validators_by_power_iter(ctx)? {
+            let validator = validator?;
+
+            curr_validators.insert(
+                validator.operator().clone(),
+                ValidatorGovInfo {
+                    address: validator.operator().clone(),
+                    bounded_tokens: validator.bonded_tokens().clone(),
+                    delegator_shares: validator.delegator_shares().clone(),
+                    delegator_deduction: Decimal256::zero(),
+                    vote: Vec::new(),
+                },
+            );
+        }
+
+        let mut tally_results = TallyResultMap::new();
+        let mut total_voting_power = Decimal256::zero();
+
+        for vote in WeightedVoteIterator::new(ctx.kv_store(&self.store_key), proposal_id)
+            .map(|this| this.map(|(_, value)| value))
+            .collect::<Vec<_>>()
+        {
+            let MsgVoteWeighted {
+                proposal_id: _,
+                voter,
+                options: vote_options,
+            } = vote?;
+
+            let val_addr = ValAddress::from(voter.clone());
+            if let Some(validator) = curr_validators.get_mut(&val_addr) {
+                validator.vote = vote_options.clone();
+            }
+
+            for delegation in self
+                .staking_keeper
+                .delegations_iter(ctx, &voter)
+                .collect::<Vec<_>>()
+            {
+                let delegation = delegation?;
+
+                if let Some(validator) = curr_validators.get_mut(delegation.validator()) {
+                    // from cosmos: https://github.com/cosmos/cosmos-sdk/blob/d3f09c222243bb3da3464969f0366330dcb977a8/x/gov/keeper/tally.go#L51
+                    // There is no need to handle the special case that validator address equal to voter address.
+                    // Because voter's voting power will tally again even if there will deduct voter's voting power from validator.
+
+                    validator.delegator_deduction += delegation.shares(); // TODO: handle overflow?
+
+                    // delegation shares * bonded / total shares
+                    let voting_power = delegation
+                        .shares()
+                        .mul(Decimal256::new(validator.bounded_tokens))
```

Comment:
It seems like here can be mistake with multiplication or with decimal. Please, check it.

Reply:
You mean replace methods call with operators or...?

Reply:
If I'm correct, you've created a number that has value `n * 10 ^ (-18)`. But seems like you need to have a decimal representation of an integer. 
the other thing is the overflows. Can you get an overflow?

Reply:
> But seems like you need to have a decimal representation of an integer.
I forget about it completely. Even that I gave you a similar comment in one of your PR's. Remind me I need `from_atomics( bounded_tokens, 0)`?
I don't handle any overflow for now